hedera-node/src/main/java/com/hedera/services/ledger/PureTransferSemanticChecks.java

```diff
@@ -0,0 +1,194 @@
+
+import java.math.BigInteger;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.ACCOUNT_REPEATED_IN_ACCOUNT_AMOUNTS;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.BATCH_SIZE_LIMIT_EXCEEDED;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.EMPTY_TOKEN_TRANSFER_ACCOUNT_AMOUNTS;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.INVALID_ACCOUNT_AMOUNTS;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.INVALID_ACCOUNT_ID;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.INVALID_TOKEN_ID;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.OK;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.TOKEN_ID_REPEATED_IN_TOKEN_LIST;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.TOKEN_TRANSFER_LIST_SIZE_LIMIT_EXCEEDED;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.TRANSFERS_NOT_ZERO_SUM_FOR_TOKEN;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.TRANSFER_LIST_SIZE_LIMIT_EXCEEDED;
+import static java.math.BigInteger.ZERO;
+
+/**
+ * Given a maximum number of ℏ and token unit balance changes in a CryptoTransfer,
+ * validates the actual transfers requested in such a transaction.
+ *
+ * Since in the normal transaction lifecycle, this logic will be performed
+ * during {@link com.hedera.services.txns.span.SpanMapManager#expandSpan(TxnAccessor)},
+ * we can accept some inefficient use of gRPC types.
+ */
+public class PureTransferSemanticChecks {
+	public ResponseCodeEnum fullPureValidation(
+			TransferList hbarAdjustsWrapper,
+			List<TokenTransferList> tokenAdjustsList,
+			ImpliedTransfersMeta.ValidationProps validationProps
+	) {
+		final var maxHbarAdjusts = validationProps.getMaxHbarAdjusts();
+		final var maxTokenAdjusts = validationProps.getMaxTokenAdjusts();
+		final var maxOwnershipChanges = validationProps.getMaxOwnershipChanges();
+
+		final var hbarAdjusts = hbarAdjustsWrapper.getAccountAmountsList();
+
+		if (hasRepeatedAccount(hbarAdjusts)) {
+			return ACCOUNT_REPEATED_IN_ACCOUNT_AMOUNTS;
+		}
+		if (!isNetZeroAdjustment(hbarAdjusts)) {
+			return INVALID_ACCOUNT_AMOUNTS;
+		}
+		if (!isAcceptableSize(hbarAdjusts, maxHbarAdjusts)) {
+			return TRANSFER_LIST_SIZE_LIMIT_EXCEEDED;
+		}
+
+		final var tokenValidity = validateTokenTransferSizes(tokenAdjustsList, maxTokenAdjusts, maxOwnershipChanges);
+		if (tokenValidity != OK) {
+			return tokenValidity;
+		}
+		return validateTokenTransferSemantics(tokenAdjustsList);
+	}
+
+	ResponseCodeEnum validateTokenTransferSizes(
+			List<TokenTransferList> tokenTransfersList,
+			int maxListLen,
+			int maxOwnershipChanges
+	) {
+		final int numScopedTransfers = tokenTransfersList.size();
+		if (numScopedTransfers == 0) {
+			return OK;
+		}
+
+		if (numScopedTransfers > maxListLen) {
```

Comment:
Ah I see what you mean...deleted the `if` on L96 so that we only return `TOKEN_TRANSFER_LIST_SIZE_LIMIT_EXCEEDED` if the count of fungible transfers exceeded `maxListLen`.